USBUS: Initial work towards an USB stack

bergzand opened the pull request:
### Contribution description

USBUS (Universal Serial Bus Unified Stack) is a generic USB stack specific for RIOT-os. It is a newly written USB stack making use of the usbdev module proposed in #9830. The idea is to have a single message/event based thread handling all the usb work. 

Modularity is achieved by allowing different endpoint modules to register themselves with their specific settings. On startup, each module requiring USB endpoints have to request endpoints from USBUS and add a callback for handling the messages. Examples of modules that could be created are: USB CDC ACM (serial over USB), USB HID (sensors over USB) and USB CDC EEM (ethernet over USB).

### Testing procedure

Run the `examples/usbus_minimal` on a `samr21-xpro` board. Plugging in the second (TARGET USB) in an USB host device (your desktop) should cause it to show up in the usb device listing with something like:
```
% lsusb  -d 046d:5678 -v

Bus 003 Device 018: ID 046d:5678 Logitech, Inc. 
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x046d Logitech, Inc.
  idProduct          0x5678 
  bcdDevice            0.00
  iManufacturer           3 
  iProduct                2 
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength            9
    bNumInterfaces          0
    bConfigurationValue     1
    iConfiguration          1 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
```
(I'm fully aware that I'm stealing one of the Logitech vendor codes, It's on my todo for #9830 to change it to something else).

### Issues/PRs references

depends on #10915 and #9830 

![image](https://bergzand.net/misc/usbus.dot.svg)

### Todo:
 - [x] auto_init (#11075)

dylad commented:
@bergzand I'll have a look this week-end or next week.
I don't see any `examples/usbus_minimal` with this PR. Is this missing or will it come in a followup PR ?

bergzand commented:
>  I don't see any `examples/usbus_minimal` with this PR. Is this missing or will it come in a followup PR ?

Must have forgotten to commit/push this, thanks for notifying me

dylad commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+                usbus->addr = 0;
+                usbus->setup_state = USBUS_SETUPRQ_READY;
+                usbus->dev->driver->set(usbus->dev, USBOPT_ADDRESS, &usbus->addr, sizeof(uint8_t));
+                puts("Reset");
```
Huh,  maybe something like usbus: event reset
Also why do you use puts() and not the DEBUG() function ?

bergzand replied:
Because of lazy probably, sorry :(

keestux commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,680 @@
+                usbdev_ep_ready(ep, 0);
+            }
+            res = sizeof(usbopt_enable_t);
+            res = sizeof(usbopt_enable_t);
```
Duplicate line?

bergzand replied:
Yes, fixing this in #10915

keestux commented on cpu/sam0_common/include/sam_usb.h:
```diff
@@ -0,0 +1,64 @@
+ */
+
+/**
+ * @defgroup    sys_uuid RFC 4122 compliant UUID's
```
Copy-paste error?

keestux commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,680 @@
+static inline void poweroff(void)
+{
+#if defined(CPU_FAM_SAMD21)
+    PM->AHBMASK.reg |= PM_AHBMASK_USB;
```
This should be clearing the bit. And also the APBBMASK is missing here. (Besides, the poweroff function is not used)

bergzand replied:
Ugh, let me check this one later, completely messed the function up

kaspar030 commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+                      .content = { .ptr = usbdev } };
+
+        if (msg_send(&msg, usbus->pid) <= 0) {
+            puts("usbus: possibly lost interrupt.");
```
I'd seriously consider using thread flags for this. You'd disable this failure case entirely, thread flags cannot get lost.

An arriving message also sets a thread flag, so a thread can wait for both:

```
thread_flag_t flags = thread_flag_wait_any(0xffff);
if (flags & THREAD_FLAG_MSG_WAITING) {
  msg_t msg
  while (msg_try_receive(&msg) == 1) {
     // handle netapi messages
  }
}
  if (flags & YOUR_FLAG_SET_BY_ISR) {
    /// handle
  }
}
```

kaspar030 replied:
I understand if you'd rather not go into this for now, can be changed later.

bergzand replied:
This is actually one of the issues where I'd really would like to have some feedback on. I've used messages here because it's what I'm most familiar with from gnrc, but using thread flags or events are both open to suggestions for me.

bergzand replied:
After some out-of-band discussion it became clear that it should be easy to move to thread flags with a flag for the generic usbdev ISR and a flag for endpoint specific ISR's. A uint32_t can keep track of the endpoints that require servicing (USB has 32 endpoints max, 16 IN and 16 OUT). This has as a consequence that usbus needs fast access to the usbdev endpoints, preferably O(1) complexity to resolve the thread flag back to an endpoint.

One solution here is to allocate the `usbus_endpoint_t` (not to be confused with `usbdev_ep_t`) as a static array. This could then reuse the `USBDEV_NUM_ENDPOINTS`, matching the number of endpoints allocated by the peripheral.

bergzand replied:
@kaspar030 I rewrote the loop to use thread flags. 2 flags are used as described above. Additionally I've added handling for `events` to allow signalling events to usb functionality from outside or inside the usbus thread.

keestux commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,680 @@
+
+void usb_attach(void)
+{
+    /* Datasheet is not clear whether device starts detached */
```
The comment does not belong here. This function is simply doing what it's been told to do.
BTW. I don't have a direct answer, but ASF usb init and Arduino bootloader do an attach. So I guess it is detached when the device starts.

bergzand replied:
This served as a reminder for myself to check this. As you said, this doesn't belong here, I'm removing it.

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,680 @@
+
+static bool usb_enable_syncing(void)
+{
+    if (USB->DEVICE.SYNCBUSY.reg & USB_SYNCBUSY_ENABLE) {
```
Agreed, fixed in #10915

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,680 @@
+
+static bool usb_swrst_syncing(void)
+{
+    if (USB->DEVICE.SYNCBUSY.reg & USB_SYNCBUSY_SWRST) {
```
Agreed, fixed in #10915

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,680 @@
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
+    MCLK->AHBMASK.reg |= (MCLK_AHBMASK_USB);
```
Legacy probably, removed

bergzand commented:
rebased to the current #10915

dylad commented:
@bergzand examples/usbus_minimal doesn't build.
```
/home/dylad/software/RIOT/examples/usbus_minimal/bin/arduino-zero/usbus.a(usbus.o): In function `usbus_init':
/home/dylad/software/RIOT/sys/usb/usbus/usbus.c:76: undefined reference to `driver'
collect2: error: ld returned 1 exit status
```

dylad commented:
@bergzand nvm, I miss the 
```
ifneq (,$(filter usbdev,$(USEMODULE)))
  USEMODULE += sam_usb
endif
```
I was able to successfully test it on arduino-zero.

keestux commented:
Not much luck with my SODAQ board, which are almost the same as Arduino Zero.
I'v added the `USEMODULE += sam_usb` in the boards Makefile and enabled debug in `usb.c` and `usbus.c`

```
main(): This is RIOT! (Version: 2018.10-devel-527-gaf3ff6-rapper.fritz.box-usb-sodaq-one)
RIOT USB stack example application
usbus: starting thread 3
usbus: Adding string descriptor number 1 for: "USB config"
usbus: Adding string descriptor number 2 for: "USB device"
usbus: Adding string descriptor number 3 for: "RIOT-os.org"
Reset
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
Unhandled out 0: 0
Reset
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
Setting addres 35
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep0: possibly lost interrupt.
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep0: possibly lost interrupt.
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep0: possibly lost interrupt.
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep0: possibly lost interrupt.
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep0: possibly lost interrupt.
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep0: possibly lost interrupt.
usbus_ep: pid: 3
usbus_ep: pid: 3
usbus_ep: pid: 3
Unhandled in 0: 8
Unhandled in 0: 8
Unhandled in 0: 8
Unhandled in 0: 8
Unhandled in 0: 8
Unhandled in 0: 8
```

keestux commented:
On the other hand, maybe I'm expecting too much at this point. What I do get is:
```
# lsusb -d 046d:5678 -v

Bus 001 Device 041: ID 046d:5678 Logitech, Inc. 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x046d Logitech, Inc.
  idProduct          0x5678 
  bcdDevice            0.00
  iManufacturer           3 (error)
  iProduct                2 (error)
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength            9
    bNumInterfaces          0
    bConfigurationValue     1
    iConfiguration          1 (error)
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
Device Status:     0x9cd0
  (Bus Powered)
  Debug Mode
```

bergzand commented:
@keestux Did you try with debugging disabled? My experience with running with debug logging is that the delays caused by the serial transfer are enough to cause timing issues with USB.

It is still on my todo list to figure out how much debugging can be enabled without causing issues

keestux commented:
@bergzand I think you're right. Without debug `lsusb` shows
```
# lsusb -d 046d:5678 -v

Bus 001 Device 043: ID 046d:5678 Logitech, Inc. 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x046d Logitech, Inc.
  idProduct          0x5678 
  bcdDevice            0.00
  iManufacturer           3 RIOT-os.org
  iProduct                2 USB device
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength            9
    bNumInterfaces          0
    bConfigurationValue     1
    iConfiguration          1 USB config
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
Device Status:     0x0000
  (Bus Powered)
```
On the RIOT device console it shows just two "Reset" messages:
```
main(): This is RIOT! (Version: 2018.10-devel-527-gaf3ff6-rapper.fritz.box-usb-sodaq-one)
RIOT USB stack example application
Reset
Reset
```

keestux commented on sys/usb/usbus/hdrs.c:
```diff
@@ -0,0 +1,231 @@
+            alt;
+            alt = alt->next) {
+        usb_descriptor_interface_t usb_iface;
+        memset(&usb_iface, 0 , sizeof(usb_descriptor_interface_t));
```
Space after `0`
It occurs in more places where there is a `memset`

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+void _req_status(usbus_t *usbus)
+{
+    uint8_t status[2];
+    memset(status, 0, 2);
```
Instead of `2` it is safer to use `sizeof(status)`

bergzand replied:
I should know this

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+    uint8_t status[2];
+    memset(status, 0, 2);
+    usbus_put_bytes(usbus, status, sizeof(status));
+    usbus->in->driver->ready(usbus->in, 2);
```
Consider `2` => `sizeof(status)` if that makes sense

bergzand replied:
Done

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+    usbus->out->driver->ready(usbus->out, 0);
+}
+
+void _req_status(usbus_t *usbus)
```
Are these `_req_xxx` functions meant to be `static`?

bergzand replied:
Yes, fixing this in the next push

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+    else {
+        usb_descriptor_string_t desc;
+        desc.type = USB_TYPE_DESCRIPTOR_STRING;
+        mutex_lock(&usbus->lock);
```
It would help the reader to know what the lock is protection...
Otherwise one might ask why there is no lock in the section above.

bergzand replied:
These locks originate from the idea to have multiple threads interacting with the stack. I don't think that this was a very good idea, so for now it is one single thread and it should be possible to remove most if not all of these locks.

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+            usbus_ep0_ready(usbus);
+        }
+        else {
+            usbus_ep0_ready(usbus);
```
Nothing has been "put". Is this OK then? To do `usbus_ep0_ready`?

dylad replied:
It is. Some times you just want to notify things without send any data.

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+    return pkt->length > usbus->in->len ? usbus->in->len : pkt->length;
+}
+
+void recv_setup(usbus_t *usbus, usbdev_ep_t *ep)
```
You probably know better than I, do we have a function naming convention? I mean, since C has a single name space, there could easily arise a name conflict with functions named like this.
If this becomes a `static` function then there is no problem, of course.

bergzand replied:
Should probably also be static

kaspar030 replied:
> do we have a function naming convention?

Yes, public (non-static) functions should be called ```module_function()``` (e.g., ```usbus_recv_setup()```). static functions should start with underline (```static void _recv_setup(void)```).

bergzand replied:
marked almost all these functions as static, please let me know if I missed one

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+                usbus->state = USBUS_STATE_RESET;
+                usbus->addr = 0;
+                usbus->setup_state = USBUS_SETUPRQ_READY;
+                usbus->dev->driver->set(usbus->dev, USBOPT_ADDRESS, &usbus->addr, sizeof(uint8_t));
```
Is the `sizeof` correct? I'm asking because `usbus->addr` is a `uint16_t`.
And if your answer is no, then I would strongly suggest to change it to `sizeof(usbus->addr)`
And if your answer is yes, you should realize you have an endianess problem.

bergzand replied:
Not that it is breaking anything at the moment, but `usbus->addr` should be `uint8_t` :confused:

keestux replied:
Even if the `addr` should be `uint8_t`, it is safer to program this as `sizeof(usbus->addr)`.

bergzand replied:
Agreed, fixed

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,667 @@
+                }
+                break;
+            case USBDEV_EVENT_TR_FAIL:
+                if (ep->dir == USB_EP_DIR_OUT) {
```
Maybe add TODO marker? That is, if something needs to be done here. Otherwise drop the `if`

bergzand replied:
Done

dylad commented on sys/include/usb.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief USB peripheral device vendor ID
+ */
+#ifndef USB_CONFIG_VID
+#define USB_CONFIG_VID          (0x046d)
```
I proposed to ask for a free PID from [here](http://pid.codes/) in #9830

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,178 @@
+#include <stdlib.h>
+#include "usb/usbdev.h"
+#include "usb.h"
+#include "usb/message.h"
```
I prefer the long form to prevent confusion for a future reader

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+    UsbDeviceEndpoint *ep_reg = &USB->DEVICE.DeviceEndpoint[ep->num];
+
+    if (ep->dir == USB_EP_DIR_OUT) {
+        ep_reg->EPINTENCLR.reg = USB_DEVICE_EPINTENCLR_TRCPT0 |
```
Ack, should be fixed

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+        res->num = 0;
+    }
+    else {
+        for (unsigned idx = 1; idx < 8; idx++) {
```
Whoops :-(

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+
+}
+
+static void _ep_size(usbdev_ep_t *ep)
```
Renamed

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+}
+
+
+static void _ep_address(usbdev_ep_t *ep)
```
renamed

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+    assert(ep != NULL);
+    switch (opt) {
+        case USBOPT_EP_ENABLE:
+            assert(value_len == sizeof(usbopt_enable_t));
```
Fixed

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+            res = sizeof(char *);
+            break;
+        case USBOPT_EP_BUF_SIZE:
+            assert(value_len == sizeof(size_t));
```
Whole setter is removed

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+            break;
+        case USBOPT_EP_BUF_ADDR:
+            assert(value_len == sizeof(uint8_t *));
+            res = sizeof(char *);
```
Whole setter is removed

bergzand commented:
rebased and reworked for the changes in #10915 and #9830.

Added some documentation (Always needs more docs, but maybe later). renamed and moved around some functions, mainly the `usbus_control` functions. Added a lot of `static`

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+void usbdev_init(usbdev_t *dev)
+{
+    /* Only one usb device on this board */
+    _usbdev = (sam0_common_usb_t *)dev;
```
Reworked this to not have a hard dependency on the `sam0_common_usb_t` struct

bergzand commented on drivers/include/usb/usbdev.h:
```diff
@@ -0,0 +1,337 @@
+ */
+typedef enum {
+    /**
+     * @brief Driver needs it's ISR handled
```
Fixed in one of the previous PR's

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+            event = USBDEV_EVENT_TR_STALL;
+        }
+        else {
+            DEBUG("Unhandled out %u: %x\n", ep->num, ep_reg->EPINTFLAG.reg);
```
fixed

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+            ep->cb(ep, USBDEV_EVENT_ESR);
+        }
+        else {
+            DEBUG("Unhandled ISR\n");
```
Reworded

bergzand commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,672 @@
+        }
+    }
+    else {
+        _disable_irq();
```
Added, also added a comment a bit above to clarify that those are endpoint interrupt cases

bergzand commented:
No longer WIP btw

keestux commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,324 @@
+void usbus_init(usbus_t *usbus, usbdev_t *usbdev)
+{
+    /* TODO: Check if memset works/is necessary */
+    memset(usbus, 0, sizeof(usbus));
```
Shouldn't this be `sizeof(*usbus)` ?

bergzand replied:
Yes

Noticed it after I rebased and pushed everything :cry:

Fixed

keestux commented:
Is this the right place for this?
```
main(): This is RIOT! (Version: 2018.10-devel-544-g6bceee-rapper.fritz.box-usb-sodaq-one)
RIOT USB stack example application
All up, running the shell now
> 0x3797
*** RIOT kernel panic:
FAILED ASSERTION.

*** halted.
```

bergzand commented:
> Is this the right place for this?

Sure, no worries. If possible next time please add `CFLAGS += -DDEBUG_ASSERT_VERBOSE` to the makefile, then the line number where the assert failure occured is printed.

I must have forgotten to actually commit and push the change that modified the `usbus::addr` member to an `uint8_t`, thus causing an assertion failure in the usbdev peripheral.

keestux commented:
Yes, much better now

bergzand commented:
rebased

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,515 @@
+     * @param event     @ref USBUS_MSG_TYPES "event" to handle
+     * @param arg       Additional argument
+     */
+    int (*event_handler)(usbus_t *usbus, struct usbus_handler *handler, uint16_t event, void *arg);
```
One flaw of the API I've found so far is that too many event-like operations are multiplexed over this single callback. Splitting this into multiple callbacks {events, transfer, setup} reduces the API complexity, allows for not handling certain events (leaving the function ptr at `NULL`) and prevents every handler from having [switch-case statements](https://github.com/RIOT-OS/RIOT/pull/11085/files#diff-4f2c0a81a745c2641ad7a3c388997c8dR313) to demultiplex the reported event

bergzand replied:
Splitted these, documentation can be improved and a few defines should be changed, but so far I'm happy with it and it reduces code size a bit

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,515 @@
+typedef struct {
+    gen_hdr get_header;         /**< function ptr to retrieve the header */
+    hdr_len get_header_len;     /**< function ptr to retrieve the header len */
+} usbus_hdr_gen_funcs_t;
```
Experience has learned that most headers have a fixed size. Changing this to:
```
typedef struct {
    gen_hdr get_header;         /**< function ptr to retrieve the header */
    union {
        hdr_len get_header_len;     /**< function ptr to retrieve the header len */
        size_t fixed;               /**< Fixed length of the header */
    } len;
    usbus_hdr_len_t len_type     /**< Either USBUS_HDR_LEN_FIXED or USBUS_HDR_LEN_FUNC */
} usbus_hdr_gen_funcs_t;
```
Allows for both fixed size and function based at almost zero cost (current function calls for the length are 4B of flash)

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,515 @@
+    size_t len;                     /**< Length of the full message       */
+    size_t transfered;              /**< Number of bytes transfered       */
+    size_t reqlen;                  /**< Maximum length of the request    */
+} usbus_controlslicer_t;
```
Should be moved to it's own header file with the function implementations in `usbus_controlslicer.c`

bergzand replied:
Shuffled things around a bit

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,515 @@
+    usbus_controlslicer_t slicer; /**< Slicer for multipart control messages */
+    usbdev_t *dev;                /**< usb phy device of the usb manager     */
+    usbdev_ep_t *out;             /**< EP0 out endpoint                      */
+    usbdev_ep_t *in;              /**< EP0 in endpoint                       */
```
Should preferably moved to `usbdev_ep0_handler_t`

bergzand replied:
moved

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,515 @@
+    const usbus_handler_driver_t *driver; /**< driver for this handler */
+    usbus_interface_t *iface;             /**< Interface this handler belongs
+                                               to */
+    uint32_t flags;                       /**< Report flags */
```
Maybe learn from [previous discussion](https://github.com/RIOT-OS/RIOT/pull/10214#discussion_r260477918) and define a `usbus_handler_flags_t` for this

dylad commented:
@bergzand Sorry to bother you again. May I ask you to rebased this on top of #10915 so I can also test on hardware ?

bergzand commented:
> @bergzand Sorry to bother you again. May I ask you to rebased this on top of #10915 so I can also test on hardware ?

Give me a few minutes, almost done making sure that my rebase is functional

bergzand commented:
@dylad Rebased. Note that because there is no longer a default VID/PID, both have to be supplied with the example makefile:
```
CFLAGS += -DUSB_CONFIG_VID=0xYOURVID -DUSB_CONFIG_PID=0xYOURPID
```

bergzand commented:
rebased again with the latest changes from #10915

bergzand commented:
No longer WIP btw

dylad commented:
> rebased again with the latest changes from #10915

Nice, thanks for doing it so quickly !

dylad commented:
@bergzand gcc is complaining 

```
$ make BOARD=arduino-zero -C examples/usbus_minimal
make : on entre dans le répertoire « /home/dylad/software/RIOT/examples/usbus_minimal »
Building application "usbus_minimal" for "arduino-zero" with MCU "samd21".

"make" -C /home/dylad/software/RIOT/boards/arduino-zero
"make" -C /home/dylad/software/RIOT/core
"make" -C /home/dylad/software/RIOT/cpu/samd21
"make" -C /home/dylad/software/RIOT/cpu/cortexm_common
"make" -C /home/dylad/software/RIOT/cpu/cortexm_common/periph
"make" -C /home/dylad/software/RIOT/cpu/sam0_common
"make" -C /home/dylad/software/RIOT/cpu/sam0_common/periph
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c: In function '_usbdev_init':
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:308:29: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init(usbdev->config.dp, GPIO_IN);
                             ^
                             ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:309:29: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init(usbdev->config.dm, GPIO_IN);
                             ^
                             ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:310:33: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init_mux(usbdev->config.dm, usbdev->config.dm_mux);
                                 ^
                                 ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:310:52: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init_mux(usbdev->config.dm, usbdev->config.dm_mux);
                                                    ^
                                                    ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:311:33: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init_mux(usbdev->config.dp, usbdev->config.dp_mux);
                                 ^
                                 ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:311:52: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init_mux(usbdev->config.dp, usbdev->config.dp_mux);
                                                    ^
                                                    ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:336:5: error: implicit declaration of function 'gpio_init_int'; did you mean 'gpio_init_mux'? [-Werror=implicit-function-declaration]
     gpio_init_int(usbdev->config.vbus, GPIO_IN_PD, GPIO_BOTH,
     ^~~~~~~~~~~~~
     gpio_init_mux
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:336:33: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     gpio_init_int(usbdev->config.vbus, GPIO_IN_PD, GPIO_BOTH,
                                 ^
                                 ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:338:33: error: 'usbdev->config' is a pointer; did you mean to use '->'?
     if (gpio_read(usbdev->config.vbus)) {
                                 ^
                                 ->
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c: In function '_usbdev_esr':
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:496:23: error: 'usbus' undeclared (first use in this function); did you mean 'usbdev'?
         if (gpio_read(usbus->config.vbus)) {
                       ^~~~~
                       usbdev
/home/dylad/software/RIOT/cpu/sam0_common/periph/usbdev.c:496:23: note: each undeclared identifier is reported only once for each function it appears in
cc1: all warnings being treated as errors
/home/dylad/software/RIOT/Makefile.base:83: recipe for target '/home/dylad/software/RIOT/examples/usbus_minimal/bin/arduino-zero/sam0_common_periph/usbdev.o' failed
```

bergzand commented:
> @bergzand gcc is complaining

Sorry, I was too quick with the fixups :cry: Should have been fixed with the latest (force) push, see [here](https://github.com/RIOT-OS/RIOT/compare/2b7cdae4062a3901711b0574778dacccb5921834..a4d40f31b8d016889c827e38c757ee9b8dcad405)

bergzand commented:
@dylad Rebased on top of #10915 

- removed the `HOST_CONNECT`/`HOST_DISCONNECT` handling, not enough boards use this to rely on this.
- added suspend/resume support.

dylad commented:
@bergzand I've got some bad news...

```
main(): This is RIOT! (Version: 2018.10-RC1-1850-g21098-pr)
RIOT USB stack example application

Context before hardfault:
```

bergzand commented:
> @bergzand I've got some bad news...

Is this on the arduino-zero? Do you have some more debugging information for me? Bit of context (stacktrace maybe)?

bergzand commented:
Hmmm, I can reproduce it on the samr21-xpro, it was working earlier this evening

bergzand commented:
Okay found it, will provide a fix tomorrow morning

dylad commented:
@bergzand examples/usbus_minimal is working on arduino-zero !

bergzand commented:
> @bergzand examples/usbus_minimal is working on arduino-zero !

:tada:

bergzand commented:
rebased to current #10915

bergzand commented:
rebased to current master

bergzand commented:
rebased again and pushed some fixups

bergzand commented:
@dylad I still want to uncrustify this and check the docs, but feel free to start with a high level review. I know that a README.md for the example is still missing.

I've reworked #11077 to adhere to the new api so that you have an actual API implementation to work with.

bergzand commented:
uncrustified!

dylad commented on sys/include/usb/usbus/control.h:
```diff
@@ -0,0 +1,92 @@
+    usbdev_ep_t *in;                    /**< EP0 in endpoint                  */
+} usbus_control_handler_t;
+
+int usbus_control_init(usbus_t *usbus, usbus_control_handler_t *handler);
```
missing doc ?

bergzand replied:
added

dylad commented on sys/include/usb/usbus/control.h:
```diff
@@ -0,0 +1,92 @@
+ */
+void usbus_control_slicer_ready(usbus_t *usbus);
+
+int usbus_control_slicer_nextslice(usbus_t *usbus);
```
missing doc

bergzand replied:
added

dylad commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,334 @@
+ */
+
+#include "thread.h"
+#include "byteorder.h"
```
needed ?

bergzand replied:
Removed

dylad commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,334 @@
+    while (1) {
+        msg_receive(&msg);
+        /* dispatch USBUS messages */
+        switch (msg.type & 0xFF00) {
```
Is the mask really needed ? If yes, please add a note for explaining black magic.

bergzand replied:
refactored out while implementing [the comment](https://github.com/RIOT-OS/RIOT/pull/10916#discussion_r253214624) from @kaspar030

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,408 @@
+    }
+}
+
+static size_t usbus_cpy_str_to_utf16(usbus_t *usbus, const char *str)
```
```suggestion
static size_t _usbus_cpy_str_to_utf16(usbus_t *usbus, const char *str)
```

bergzand replied:
Fixed

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,408 @@
+        /* TODO: implement device qualifier support (only required
+         * for High speed) */
+    }
+    return -1;
```
Why -1 ??

bergzand replied:
Added a comment (negative signals a stall condition, [see also here](https://github.com/RIOT-OS/RIOT/pull/10916/files#diff-38e020bc78b287d1782f6e35d81afcefR329))

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,408 @@
+            return _req_str(usbus, idx);
+        case USB_TYPE_DESCRIPTOR_DEV_QUALIFIER:
+            return _req_dev_qualifier(usbus);
+        default:
```
please add a DEBUG statement here and in other default statements below.

bergzand replied:
I think I got them all, but feel free to point out if I missed some

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,408 @@
+                               usbus_control_handler_t *ep0_handler,
+                               usbdev_ep_t *ep)
+{
+    switch (ep0_handler->setup_state) {
```
missing default statement.

bergzand replied:
added

dylad commented on sys/usb/usbus/usbus_fmt.c:
```diff
@@ -0,0 +1,279 @@
+
+static uint8_t _type_to_attribute(usbus_endpoint_t *ep)
+{
+    switch (ep->ep->type) {
```
missing default statement

bergzand replied:
Added

dylad commented on sys/usb/usbus/usbus_fmt.c:
```diff
@@ -0,0 +1,279 @@
+    return len;
+}
+
+size_t _alt_size(usbus_t *usbus, usbus_interface_alt_t *alt)
```
should be ?
```suggestion
static size_t _alt_size(usbus_t *usbus, usbus_interface_alt_t *alt)
```

bergzand replied:
Done

dylad commented on sys/usb/usbus/usbus_fmt.c:
```diff
@@ -0,0 +1,279 @@
+    return _hdrs_fmt_additional(usbus, usbus->hdr_gen);
+}
+
+size_t _hdrs_fmt_endpoints(usbus_t *usbus, usbus_endpoint_t *ep)
```
```suggestion
static size_t _hdrs_fmt_endpoints(usbus_t *usbus, usbus_endpoint_t *ep)
```

bergzand replied:
Done

dylad commented on sys/usb/usbus/usbus_fmt.c:
```diff
@@ -0,0 +1,279 @@
+        conf.attributes |= USB_CONF_ATTR_SELF_POWERED;
+    }
+    /* Todo: upper bound */
+    conf.max_power = USB_CONFIG_MAX_POWER / 2;
```
why / 2 ?

bergzand replied:
the bMaxPower (table 9-10 from the usb 2.0 spec) is in 2mA increments. I prefer to have the define in actual mA for convenience.

bergzand replied:
I can put a comment here to explain

dylad replied:
I would be very happy !

bergzand commented on sys/Makefile:
```diff
@@ -139,6 +139,9 @@ endif
 ifneq (,$(filter bluetil_%,$(USEMODULE)))
   DIRS += net/ble/bluetil
 endif
+ifneq (,$(filter usbus usbus_%,$(USEMODULE)))
```
It's a bit of future planning. I can remove it here but I would have to include it again in one of the follow ups.

bergzand commented on sys/include/usb/descriptor.h:
```diff
@@ -0,0 +1,223 @@
+#define USB_TYPE_DESCRIPTOR_DEV_QUALIFIER   0x06    /**< Device qualifier          */
+#define USB_TYPE_DESCRIPTOR_SPEED_CONFIG    0x07    /**< Other speed configuration */
+#define USB_TYPE_DESCRIPTOR_IFACE_POWER     0x08    /**< Interface power           */
+#define USB_TYPE_DESCRIPTOR_INTERFACE_ASSOC 0x0B    /**< Interface association     */
```
Section 9.X.Y, table 9-Z, not kidding :laughing: 

See the USB engineering change notice (ECN) about interface association descriptors in the usb 2.0 spec zip archive.

bergzand replied:
Changed :blush:

bergzand commented on sys/include/usb/descriptor.h:
```diff
@@ -0,0 +1,223 @@
+#define USB_SETUP_REQ_GET_CONFIGURATION     0x08    /**< Get configuration     */
+#define USB_SETUP_REQ_SET_CONFIGURATION     0x09    /**< Set configuration     */
+#define USB_SETUP_REQ_GET_INTERFACE         0x0a    /**< Get interface         */
+#define USB_SETUP_REQ_SET_INTERFACE         0x0b    /**< Set interface         */
```
Unified

bergzand commented:
@dylad I think I've addressed all your remarks. I also did a bit of refactoring to clean up the code and finally implemented the suggestion from @kaspar030 to use thread flags and remove the `msg` code.

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,533 @@
+    usbus_string_t product;                         /**< Product string                        */
+    usbus_string_t config;                          /**< Configuration string                  */
+    usbus_endpoint_t ep_out[USBDEV_NUM_ENDPOINTS];  /**< USBUS OUT endpoints                   */
+    usbus_endpoint_t ep_in[USBDEV_NUM_ENDPOINTS];   /**< USBUS IN endpoints                    */
```
One "big" change with the latest fixup is that the `usbus_endpoint_t`'s are now statically allocated within the usbus context struct. This effectively moves them from each handler to the central context. A number equal to the usbdev endpoints is allocated, effectively matching them.

bergzand replied:
The motivation for this change lies within the change to thread flags. As a thread flag doesn't carry any content, a different way to grab the endpoint that requires servicing was necessary. The solution is a `uint32_t` acting as a holder for bitflags, one for each endpoint (USB has max 16 IN and 16 OUT endpoints as protocol limitation). From this bitmap, the endpoint direction and number that requires servicing is retrieved. The usbus endpoint context and from there the usbdev endpoint context is then retrieved from this array.

A different solution more in line with the previous decentral endpoint context storage would be to iterate all interfaces and for each interface check if it contains the endpoint that we're looking for. As this requires a lot of iterating, I decided that centralizing this storage and potentially allocated too many endpoint structs is the more favorable solution here.

dylad commented on sys/usb/usbus/usbus.c:
```diff
@@ -0,0 +1,360 @@
+            while (events) {
+                unsigned num = bitarithm_lsb(events);
+                events &= ~(1 << num);
+                if (num >= 0x10) {
```
You should explain the value.

bergzand replied:
Reworked a bit, should be explained now.

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,417 @@
+            ep0_handler->setup_state = USBUS_SETUPRQ_READY;
+            _usbus_config_ep0(ep0_handler);
+            break;
+        default:
```
missing assert/debug.

bergzand replied:
In this case the `default` catches any `usbus_event_usb_t` that are not handled by the endpoint code. I'm still thinking if an assert should be triggered in that case, but it doesn't have to indicate a failure, only that the endpoint 0 code doesn't care about the event

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,417 @@
+        case USBUS_EVENT_TRANSFER_COMPLETE:
+            _handle_tr_complete(usbus, ep0_handler, ep);
+            break;
+        default:
```
missing default assert/debug.

bergzand replied:
Same as above, but now for endpoint events instead of usb events

dylad commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,533 @@
+} usbus_handler_driver_t;
+
+/**
+ * @ USBUS handler struct
```
missing brief ?

bergzand replied:
fixed

dylad commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,533 @@
+ *
+ * @param[in] usbus USBUS context
+ * @param[in] iface USB interface to add the endpoint to
+ * @param[in] ep    USB endpoint struct to use
```
This param is not used by the function, should be removed

bergzand replied:
removed

dylad commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,533 @@
+ * @param[in] dir   USB endpoint direction
+ * @param[in] len   Buffer space for the endpoint to allocate
+ */
+usbus_endpoint_t *usbus_add_endpoint(usbus_t *usbus, usbus_interface_t *iface,
```
missing return in doxygen description

bergzand replied:
added

dylad commented on sys/usb/usbus/usbus_control.c:
```diff
@@ -0,0 +1,417 @@
+            _cpy_str_to_utf16(usbus, str->str);
+        }
+    }
+    return 1;
```
Is it needed to return 1 all the time ?

bergzand replied:
No, I should fix the function to return something different when the requested string is not found :/

bergzand replied:
Fixed function to also be able to  return `-1` in some cases

bergzand commented on sys/include/usb/descriptor.h:
```diff
@@ -0,0 +1,223 @@
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t length;             /**< Size of this descriptor                                 */
+    uint8_t type;               /**< Descriptor type (@ref USB_TYPE_DESCRIPTOR_CONFIGURATION */
```
Fixed

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,533 @@
+ *
+ * @return      Length of the generated header
+ */
+typedef size_t (*gen_hdr)(usbus_t *usbus, void *arg);
```
Reworked a bit to include the prototypes directly in the struct, please let me know if the current situation is an improvement

bergzand commented on sys/include/usb/usbus.h:
```diff
@@ -0,0 +1,533 @@
+     * @returns         Size of the returned data when the request is handled
+     * @returns         negative to have the stack return an USB stall to the
+     *                  host
+     * @returns         zero when the request is not handled by this handler
```
fixed fox consistency

bergzand commented:
@dylad I think I've fixed/answered all your comments so far.

dylad commented:
This PR is in a good shape now !
@kaspar030 @keestux Mind do have a last look and check if your comments were addressed ?
This can be merge pretty soon I think.

bergzand commented:
>  Also, there is no available command from the console with `examples/usbus_minimal` (except `help`). Does the shell is really needed ?

I forgot about this one, I guess it can be removed, maybe have the example print some USB status messages or something. It can always be extended with shell integration in a follow up

dylad commented:
@bergzand 
Do we have a solution for the default VID/PID ? Or should we let the user defines its own values ?
Otherwise I retested the whole PR, it works like a charm.

bergzand commented:
>  Do we have a solution for the default VID/PID ? Or should we let the user defines its own values ?

At the moment this is up to the user. Most free VID/PID (at least pid.codes) require a working USB stack before they want to allocate a PID, so this is a bit of a chicken&egg issue.

bergzand commented:
>  Also, there is no available command from the console with `examples/usbus_minimal` (except `help`). Does the shell is really needed ?

Reworked to only start the usb stack. I've dropped the shell, maybe we can add it back later together with some USB status commands.

miri64 commented:
Greatest fixup commit history I've ever seen xD

bergzand commented:
Squashed. Added a small `README.md` to the example

bergzand commented:
And rebased

bergzand commented:
@kaspar030 Do you have any inspiration on how to fix the build failure? To be honest I really don't want to hardcode the pid.codes test PID in the example.

bergzand commented:
Now depends on #11633 :'(

dylad commented:
Please squash and rebase

bergzand commented:
> Please squash and rebase

Done!

dylad commented:
For the record:
log from lsusb -vvv shows that VID/PID are inverted
```
Bus 001 Device 066: ID 0001:1209 Fry's Electronics 
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x0001 Fry's Electronics
  idProduct          0x1209 
  bcdDevice            0.00
  iManufacturer           3 
  iProduct                2 
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength            9
    bNumInterfaces          0
    bConfigurationValue     1
    iConfiguration          1 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA

```

bergzand commented:
>  log from lsusb -vvv shows that VID/PID are inverted

Aaah, thanks for testing, I switched around the VID/PID in the Makefile. Should be fixed now

dylad commented:
Here we go !
Awesome job from @bergzand there ! You definitely deserve a medal :)

bergzand commented:
:tada: 

@dylad, @keestux and @kaspar030 Thanks for reviewing!

emmanuelsearch commented:
@bergzand yay!

kaspar030 commented:
@bergzand niiiice!

brookstalley commented:
@bergzand thank you so much. I have been checking this thread every morning the way some people check the lottery numbers.
